### ISSUE TITLE
[Feature Request] Dynamic RPC Address Resolution

### ISSUE THREAD

**Issue by make1980** [tensorflow/tensorflow#18880]
Currently TensorFlow cluster is created statically with host name/port information defined in cluster spec. The host name and port information for a worker can't be changed during run-time. In a multi-tenancy environment where host name/port is assigned to worker by a job manager after worker failover happens, this static cluster spec prevents other workers to communicate with this new worker using the latest host name and port.

With dynamic RPC address resolution feature, the host name of each TF server is resolved against an RPC address registry before initiating each RPC call. This allows changing the host name and port of each TF server at run time, which is useful in the following cases:

- A stateless TF worker fails over and gets assigned with new host name/port by an external job scheduler.

- A stateful TF worker fails over and all workers get restarted without changing the cluster spec. The failed worker can be assigned with new host name/port.

I have composed a design document [here](https://docs.google.com/document/d/1Fl-mMZ9NMBDaQkhhVpXe-UdYC-iYg6hlSj7bBZSMCSM/edit?usp=sharing). A prototype has been done as well based on this design. We hope to contribute this feature to the upstream. Any feedbacks for the design are welcome.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by karmel** [comment]
Thank you, @make1980. @josh11b and @alextp , you each touch this space; thoughts?

**Comment by josh11b** [comment]
@saeta I think knows more about this than I do.

**Comment by josh11b** [comment]
@anj-s , FYI

**Comment by make1980** [comment]
@saeta  @alextp @anj-s @josh11b Any comments for this proposal? Would appreciate your early feedback - we plan to submit a PR soon.

**Comment by saeta** [comment]
Hey @make1980 ,

Are you familiar with ClusterSpec propagation? [ClusterSpec propagation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/client/session_clusterspec_prop_test.py) in conjunction with [ClusterResolvers](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cluster_resolver/python/training/cluster_resolver.py) allows for a fully pluggable dynamic resolution of all the workers. 

Hope this helps!
-Brennan

**Comment by make1980** [comment]
Hi @saeta ,

Thanks for the links. I looked at the implementation of ClusterSpec propagation and ClusterResolver which allows the following scenario if I understand correctly - a worker which initiated the session failed over, it can restart and construct a new cluster spec using a pluggable cluster resolver to restart a new session. ClusterSpec propagation allows all the workers participating in the new session to retrieve the IP/port of other workers (including the worker which just failed over) from the master. 

However, in this scenario for the sessions initiated from other workers, this failed-over worker can't rejoin the existing session on behalf of its previous incarnation(for example, in-graph replication case).

Also using cluster spec propagation we'll have to give the job a different name so that GRPC channel cache and worker cache won't reuse the existing cached entry based on the logical device name. This complicates the fail-over logic from the job manager's perspective since job name and logical device name should ideally remain the same when a single worker failed over.

There are definitely more details that need to be figured out such as how failed-over workers rejoining an existing session would be able to recover its local state, and how failed-over ps can restore its variables. But as the first step, I think having a translation layer from a logical hostname to IP/port at runtime is more flexible and necessary.

ClusterSpec propagation can be used to enable scenarios such as auto-scaling but we need this dynamic RPC resolution feature to easily support the single worker fail-over scenario based on my understanding.

**Comment by make1980** [comment]
To give a real-case scenario where ClusterSpec propagation isn't able to handle currently and won't be easily extended to handle -

In a 1 worker 1 ps setup, we start the worker first using cluster spec propagation. Then the machine which is supposed to run ps failed and ps ended up running on a different machine with a different IP. In this case the master service in the worker can't recognize the new ps since cluster resolver has already translated the IP/port information for the ps. To support this scenario we'll have to add logic to retry the cluster spec resolution logic (which is actually code in the user land if I understand correctly) instead of just adding retry logic where the master probes all the workers.

**Comment by make1980** [comment]
Any comments @saeta? This feature is very important for large scale training scenario where cluster manager assigns different IP/port to a worker after fail-over and we really hope to contribute this to the community to avoid conflict in core when merging with the upstream, as well as allowing more customized implementations for dynamic address resolution.

Moreover this feature should be able to work with cluster spec propagation/cluster resolver seamlessly since it provides address translation at a different layer.

**Comment by make1980** [comment]
@saeta @alextp @anj-s @josh11b Any further comments? Let me know if I missed anything here.

**Comment by make1980** [comment]
@saeta Could you comment on how to support this case using cluster spec propagation/cluster resolver?

In a 1 worker 1 ps setup, we start the worker first using cluster spec propagation. Then the machine which is supposed to run ps failed and ps ended up running on a different machine with a different IP. In this case the master service in the worker can't recognize the new ps since cluster resolver has already translated the IP/port information for the ps.

**Comment by make1980** [comment]
Any comments?

**Comment by ajbouh** [comment]
Have you thought about just using DNS for the hostname instead of an IP? You can ensure that new hosts are mapped to the same failed hostname without needing to change anything in the TensorFlow codebase. If you're worried about port numbers, you might consider using DNS srv records and writing a custom ClusterResolver.

**Comment by saeta** [comment]
Sorry @make1980 for the delayed reply. This thread got lost in my inbox. I apologize!

In general whenever there is a machine failure within a TensorFlow cluster, we have to be extra careful. This is because ops can be stateful. In short given a node failure at an arbitrary point after session creation, it's not correct to just re-try running those ops. Concretely in your scenario about 1 worker and 1 ps, if the parameter server fails after the variables have been initialized, it is not correct to begin executing the python training loop once the ps server comes back online. (Because the variables that are placed on it have not been initialized.) Even if we special-cased variables, there are many other stateful ops that also would need explicit handling. Because TensorFlow users can define custom operations, it's not possible to teach the C++-runtime how to automatically handle the failures within a session.

As a result, the failure handling strategy we adopt within TensorFlow is to close the session and create a new session (and optionally restoring from a checkpoint). When starting the new session, client code (e.g. python) can construct a new ClusterSpec and propagate it to all the nodes within the cluster. This also has the advantage of cleanly separating the core TensorFlow runtime from having to integrate with custom cluster-coordination systems (e.g. etcd, zookeeper, internal Google systems, Netflix's Eureka, k8s, etc.).

Hope this helps!

**Comment by make1980** [comment]
Thanks for your response @saeta! This is exactly the discussion that I wanted to have. As you mentioned adding run-time address resolution will cause problem for stateful operations so we avoid it by recreating the session. While this also means we might need to add automatic state recovery mechanism to PS so this won't be a problem. Checkpoint based state recovery has its own limitation of requiring all the PS servers to restore the state in case there is only one PS failing. If ultimately this can be address, dynamic RPC resolution should still be applicable. Certainly we need to consider other cases like operations relying on local state can only be resumed if the intermediate local state can be recovered. 

So I agree with you that this feature won't solve all the single-node failure problem and could be error-prone if not used properly. But it should still be useful in case the failed worker doesn't have intermediate state. For example, in an in-graph replication case where worker 1 fails before worker 2 starts the session and after worker 2 constructed the cluster spec, worker 2 doesn't need to restart the session to recognize a new worker 1. 

I understand that this is a special case where RPC resolution can bring some advantages compared with cluster spec propagation. However from the user's perspective having a consistent logical cluster view also seems to be simpler than defining different cluster specs among different workers and also using sparse cluster definition to avoid unrelated workers to wait for each other. Having a single cluster spec also provides potential opportunities for optimizations on op placement since TF can just treat all the workers in a holistic view.

Also - if worker state is the only concern for this feature request we can work on adding additional logic to avoid using this feature when the failed-over worker contains unrecoverable state. Just want to make sure the feature itself actually brings some value and we can invest more towards this direction.

**Comment by make1980** [comment]
@ajbouh, thanks for your suggestion. I took a look at the DNS srv records and I agree that this can be used to solve some of the problems we're facing together with cluster resolver. I think the difference between dynamic RCP address resolution and cluster spec propagation is really around when the address translation happens instead of how the address is translated - whether through DNS or not we can do it in both approaches. For example, we can implement an RPC address resolver using DNS srv record as well.

**Comment by ajbouh** [comment]
Perhaps a simple solution would be to actually only bring up the parts of
the session state that are stateful and affected by the failed worker?

We can see the subgraph of stateful/nonstateful operations pretty plainly
in the graphdef. Perhaps there's a simple type of op we can introduce that
returns true or false if running on a replaced worker or not. Combining
this with a cond operator and some graph rewriting should be enough to
"automatically" handle failures with the least disruption possible.

Do I have that right?

**Comment by make1980** [comment]
I don't think I fully understand your idea here - a simple solution I can imagine would be to let the op declare whether it's stateful or not and only allow a failed-over worker which has only executed stateless op to rejoin the session. And restart the session if it's not the case.

**Comment by ajbouh** [comment]
We're in luck there as operations are already marked as being
stateless/stateful for the graph optimizers to maintain correctness as they
simplify the graph.

My suggestion would make it possible to write custom (in graph) logic to
handle the stateful case as well.

**Comment by make1980** [comment]
@saeta Would appreciate if you can give some feedback on my comments above. I think address translation at run-time will still bring great value - we just need to make sure it only happens when appropriate.

The design already makes the feature an opt-in feature - we can add additional check to make sure the feature is only enabled for the right worker.

**Comment by saeta** [comment]
Hmm, this is a very nuanced topic that deserves careful thought. I guess I see 2 approaches to failure handling: either (1) we adopt the "crash-only" philosophy, or (2) we attempt some sort of "online" recovery to handle partial (e.g. single node) failures. Unfortunately, they don't mix terribly well.

Advantages of the first is that it dramatically simplifies the semantics as well as the system implementation. Because fewer code paths are required (and they are exercised constantly), software projects with this fault tolerance model can sometimes be more robust.

That said, in some domains this approach cannot meet certain system goals, and thus partial failures must be tolerated and handled correctly. (e.g. A distributed database.) Unfortunately, these systems are very complicated and are generally designed around handling failures. These systems have uptime measured over months / years and are distributed across hundreds, thousands, or more servers. By contrast, TensorFlow is instead focused on training jobs that last hours to days (weeks at most), and typically don't scale beyond hundreds of servers. Because of these differing requirements, TensorFlow has historically taken the crash-only approach. (Note: TensorFlow has actually been moving in the opposite direction; new distribution strategies centered around all-reduce and other HPC-style primitives are known to be higher performance (e.g. than parameter-server style configurations) but are easier to ensure they are working reliably in crash-only systems. Parameter-server style configurations running in an async approach are pretty resilient to workers going down.)

If we are to deviate from this, one way to think about it would be to answer the following questions:
 1. What are the overall system goals? What application are we trying to meet that we can't meet today?
 2. What is the end-to-end design? Are certain layers crash-only? If so, how does that mix with online recovery of other layers?
 3. What is the expected implementation cost of the approach? Does it prevent easy implementation of other features on our roadmap?
 4. Does the value of the applications outweigh the above costs?

Dropping down a level (if the above response is focused on the larger discussion, this is on a more detailed technical level), the cost of creating a new session is typically pretty cheap (assuming the graph is not gigantic). State can be stored in containers outside of specific sessions, so [some] more complicated partial-failure recovery can be implemented (e.g. in Python) on top of the core TF primitives today.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by saeta** [comment]
@ajbouh You're asking some excellent questions!

Unfortunately, I'm not sure that detecting worker failure can be done as an op. Ops (by design) are not allowed to know of each other, and only communicate through input and output tensors. So, if we're trying to detect failures at the (arbitrary) op level, we can't necessarily do it as a custom op. But, that's probably too hard a problem; let's scope down to just detecting worker failures, which we could detect by having an op right after the recv op (which handles the RPC communication between workers). If we wanted to go down this route, we could augment the graph re-write pass that adds in the send's and recv's to also add in our custom loaded failure detecting op. But, then two issues arise:

(1) how does the custom failure detector op know when a failure has happened? A standard timeout? If so, what's too long? It's not clear to me, given the huge variety of workloads running in TF. Even within a workload, some recv's should complete within seconds, whereas others are expected to take many minutes or even longer.

(2) Assuming we've solved (1), we now have an op (which runs on a worker) that knows a failure has occurred. The next question becomes what can be done about the failure. We'd then need to define an RPC protocol for the worker to either (a) communicate with the master to request a new graph to be executed on the failed worker to re-initialize it, or (b) have the (potentially many) workers directly re-initialize it after it comes back up, and the associated coordination required. Because RPC protocols can't be added to a gRPC server after it's been started, this would require some non-trivial re-architecting of the RPC stack. (And while this is doable, the question becomes what's the value? How many workloads can we unlock with this capability? Note that the capability itself does not solve the problem; it only allows others to write subtle and complicated C++ ops that could attempt to perform failure handling. Would any of these be sharable between TF users? Between different workloads?)

As an alternative, we could augment the TF Master to allow some custom failure handling instead of as an op. But unfortunately, it doesn't necessarily solve all the problems either (what's the expected time a worker should wait for an recv?), and introduces new challenges (e.g. we currently have no path for dynamically loading plugins into the master currently). Again, all of these are solveable with sufficient engineering resources. But there's an opportunity cost (in addition to the maintenance cost), so we have to understand the value. Given that the MTBF of machines (especially Google Compute Engine CPU VMs) is so high (a year or more), even a large cluster training for a couple days is not likely to experience a failure in the normal case. The question thus becomes, what workloads can we not serve today?

Does that make sense?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ajbouh** [comment]
Brennan, there are many ways to detect failure that should be adequate to
improve the situation. These criteria are orthogonal to the rest of the
design. Let us assume for now that upon "failure" a worker cleanly shuts
down its TCP connections and rejects any future connection attempts.

I assume that any waiting receive will raise an error of some kind. It
might be enough to modifying the receive op to *not* raise this error and
*instead* evaluate some second tensor and return 2 outputs (1 output for
the error text/type, 1 output for the second tensor).

As for how frequently this would be needed, the dropping price for hardware
and the availability of *preemptible* hardware are two trends that will
generate unique training scenarios that push the limits of today's training
architectures much more than the MBTF of underlying hardware. The design
outlined above is an initial step in that direction.

TensorFlow is uniquely suited to enable these modes. I hope we succeed in
doing just that!

Thanks for the critical feedback! :)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ewilderj** [comment]
If this work is moving forward, @ajbouh, sounds like it might make a good TensorFlow RFC, with @saeta as the sponsor. For more, see https://github.com/tensorflow/community/blob/master/governance/TF-RFCs.md

**Comment by ajbouh** [comment]
Thanks for the bump! It doesn't seem like @saeta has much buy-in.

For now I'm focused on working around this in my own projects by building
proxies, libraries, and other components that make it easier to move
quickly.

Happy to contribute to any efforts by the community though!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 91 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by saeta** [comment]
Closing this issue out as we haven't made progress in quite some time. Thanks all! -Brennan

**Comment by byronyi** [comment]
@saeta Somehow I think this topic might worth revisiting with the new eager runtime. In remote eager client like the VM connect to remote GPU or CloudTPU, one does `connect_to_cluster` and place the default context on the remote worker. If the eager service crashes, it might not be easy for the user to recover by re-running their computations which could be notebook cells and evaluated without explicit checkpoints. Any chance we could support eager context switch or checkpoint on the service side that is transparent to the client?

**Comment by alextp** [comment]
@bramandia FYI

**Comment by yuefengz** [comment]
@byronyi Launching jobs on multiple remote GPU workers is not supported. When we support it, probably via single-client architecture, checkpoints can be made periodically by the client. If any worker fails, there is already some mechanism to update the configuration of the cluster, clear eager context, rebuild resources, load from checkpoint etc. but there are other components missing to make remote task dispatching work properly. We will considering supporting that. cc @crccw
